README.md
<details>
  <summary>Abbreviations <sup>(toggle)</sup></summary>
  <pre>
  | Abbreviation | note          | Meaning                                    |
  | ---          | ---           | ---                                        |
  | "OMEMO"      | (with quotes) | refers to the official "OMEMO" plugin.     |
  | "OMEMO Mod"  | (with quotes) | refers to the present modified version.    |
  | OMEMO        | acronym       | refers to the XMPP Extension Protocol.     |
  | (ext)        |               | external hyperlink (immediately preceding) |
  | ref          |               | reference                                  |
  | DEV          |               | software development                       |
  | SecOps       |               | Security Operations [1]                    |
  </pre>
  <sup>[1] https://www.vmware.com/topics/glossary/content/secops</sup>
</details>

---


# "OMEMO Mod" plugin for Gajim

A modified version of the official OMEMO plugin for Gajim that exposes the "unacknowledged message count" (`UNACKNOWLEDGED_COUNT`) and the rest of the `backend.utils` constants as user configurable values. It works as a drop-in replacement.

These additional settings can be found in the "Advanced" section under the "Settings" tab of the "OMEMO configuration" panel/window.

Homepage: https://github.com/Gregory-K/gajim-omemo-mod

**Warning.** The modifications made to the "OMEMO" plugin are not in any way endorsed by the Gajim Development team. On the contrary, they are considered a **security flaw**. Use "OMEMO Mod" only after careful consideration (see [Reasoning](#Reasoning)). If you aren't absolutely certain you need it, please, prefer the [official one](https://dev.gajim.org/gajim/gajim-plugins/wikis/OmemoGajimPlugin)<sup>(ext)</sup>.

*ref:* [Gajim](https://gajim.org)<sup>(ext)</sup> official "OMEMO" plugin [wiki](https://dev.gajim.org/gajim/gajim-plugins/wikis/OmemoGajimPlugin)<sup>(ext)</sup> and [repository](https://dev.gajim.org/gajim/gajim-plugins)<sup>(ext)</sup>

>Quote: OMEMO Wikipedia [entry](https://en.wikipedia.org/wiki/OMEMO)<sup>(ext)</sup>  
>
>OMEMO is an extension to the Extensible Messaging and Presence Protocol ([XMPP](https://en.wikipedia.org/wiki/XMPP "XMPP")<sup>(ext)</sup>) for multi-client [end-to-end encryption](https://en.wikipedia.org/wiki/End-to-end_encryption "End-to-end encryption")<sup>(ext)</sup> developed by Andreas Straub. According to Straub, OMEMO uses the [Double Ratchet Algorithm](https://en.wikipedia.org/wiki/Double_Ratchet_Algorithm "Double Ratchet Algorithm")<sup>(ext)</sup> (Axolotl Ratchet protocol) "to provide multi-end to multi-end encryption, allowing messages to be synchronized securely across multiple clients, even if some of them are offline".  


## Current DEV State

For the time being, the "OMEMO Mod" works as intended, but it's still just a **draft hack** that covers personal needs. It does nothing more or less than its description and it doesn't "touch" any other part of the "OMEMO" source code.

**Present Branch / Release**

The `expose_all` branch exposes all the hardcoded values (constants) as user configurable values. _(Gajim v1.3)_

The `gajim_1.3` branch of the [gajim-plugins](https://dev.gajim.org/gajim/gajim-plugins)<sup>(ext)</sup> repository is being used as source code, and the following versions of Gajim are **supported**:  
. min Gajim version: 1.2.91  
. max Gajim version: 1.3.90

**Other Branches / Releases**

`Master` and `expose_msg_count` branches are the same. They expose only the "unacknowledged message count" (`UNACKNOWLEDGED_COUNT`) as a user configurable value. _(Gajim v1.3)_

The `gajim_1.4_dev` branch follows the paradigm of `expose_all`, and exposes all the hardcoded constants as user configurable values for the comming-up v1.4 of Gajim.


## Installation

Information on how to [install](#Procedure) or [update](#Updating-Procedure) this plugin.

### Auxiliary Information

This altered plugin shares the same manifest `short_name` with the original for the sole purpose of aiding the user in ensuring that only one OMEMO plugin will be loaded.

Its version will always be one or more decimal places larger than the official one, to secure that it will be loaded instead of the Gajim's "installed by default" one.  
*e.g. v2.7.14.2.1 = "OMEMO" v2.7.14 + "OMEMO Mod" v2.1*

**Attention.** Keep in mind that the Gajim "Plugin Installer" cannot auto-update the "OMEMO Mod". In case of a newer official version, since both plugins share the same manifest `short_name`, it will revert-update to the official one. "OMEMO Mod" can only be updated manually, as per the [directions](#Updating-Procedure) bellow.

"OMEMO Mod" saves its extra settings in a [pickle](https://docs.python.org/3/library/pickle.html)<sup>(ext)</sup> file named `omemo_mod.pickle` located in the `pluginsconfig` directory. The "OMEMO" SQLite DB could used, but for now, it's decided to keep the least possible modifications to the "OMEMO" source code.

#### Common Paths

`pluginsconfig` **Path**

| OS          | `pluginsconfig` Path                                      |
| ---         | ---                                                       |
| **Linux**   | `~/.config/gajim/pluginsconfig`                           |
| **Windows** | `C:\Users\<USERNAME>\AppData\Roaming\Gajim\Pluginsconfig` |
| *Any OS*    | `<custom profile path>/pluginsconfig` <sup>[1]</sup>      |

`omemo` **Path**

| OS          | `omemo` Path                                              |
| ---         | ---                                                       |
| **Linux**   | `~/.local/share/gajim/plugins/omemo`                      |
| **Windows** | `C:\Users\<USERNAME>\AppData\Roaming\Gajim\Plugins\omemo` |
| *Any OS*    | `<custom profile path>/plugins/omemo` <sup>[1]</sup>      |

<sub>[1] you custom profile path, if you 're using `Gajim -c <custom profile path>`</sub>

### Procedure

- Backup the previous "OMEMO" plugin. *(optional)*  
  e.g. Create a compressed copy of the `omemo` folder (tar / zip / 7z / etc.).
- Delete all files under the `omemo` path.

AND

- **Easy** : Download the latest [Release](https://github.com/Gregory-K/gajim-omemo-mod/releases) `.tar.gz`/`.7z` file for your Gajim version and extract it into the `omemo` path.

OR

- **Advanced** : Clone [this repository](https://github.com/Gregory-K/gajim-omemo-mod) directly into (or clone elsewhere and copy to) the `omemo` path.

### Updating Procedure

- Backup the previous "OMEMO Mod" plugin. *(optional)*  
  e.g. Create a compressed copy of the `omemo` folder (tar / zip / 7z / etc.).
- Delete all files under the `omemo` path.
- In case the "Advanced" settings storage mechanism changes (Release description will contain a warning):
  - Delete `omemo_mod.pickle` located in the `pluginsconfig` path and
  - re-apply your customized settings.

AND

- **Easy** : Download the latest [Release](https://github.com/Gregory-K/gajim-omemo-mod/releases) `.tar.gz`/`.7z` file and extract it into the `omemo` path.

OR

- **Advanced** : Checkout [this repository](https://github.com/Gregory-K/gajim-omemo-mod) directly into (or checkout to your previously chosen location and copy to) the `omemo` path.


## Configuration

Tips on tinkering with the "Advanced" settings of "OMEMO Mod".

**UNACKNOWLEDGED_COUNT**  
*number of unacknowledged messages before a device gets marked as inactive*  
Default: 300  
Allowed: 1 - 9999  
Guide Note: As per usage patterns and SecOps demands.

**DEFAULT_PREKEY_AMOUNT**  
*number of PreKeys to provide in the bundle*  
Default: 100  
Allowed: 80 - 999  
Guide Note: The bundle SHOULD always contain around 100 PreKeys.

**MIN_PREKEY_AMOUNT**  
*minimum number of PreKeys to provide in the bundle*  
Default: 80  
Allowed: 25 - 999  
Guide Note: The bundle MUST always contain at least 25 PreKeys.

**SPK_ARCHIVE_TIME**  
*time period in seconds to keep Singed PreKeys*  
Default: 1296000 seconds = 15 days  
Allowed: 86400 - 5259600 seconds (1 day to 2 months)  
Guide Note: All existing Signed PreKeys that are older than SPK_ARCHIVE_TIME will be deleted.

**SPK_CYCLE_TIME**  
*signed PreKey rotation period*  
Default: 86400 seconds = 24 hours  
Allowed: 86400 - 2629800 seconds (1 day to 1 month)  
Guide Note: Signed PreKeys SHOULD be rotated periodically once a week to once a month. A faster or slower rotation period should not be required.


## Reasoning

Justification for providing the end user with the ability to change the "OMEMO" `UNACKNOWLEDGED_COUNT` ("unacknowledged message count") value.

The ability to change the "unacknowledged message count" value is the main objective of the "OMEMO Mod". The ability to alter the rest of the "Advanced" settings is just for tinkering around and other niche cases.

### Preface

Related reading:  
- [Pitfalls for OMEMO Implementations – Part 1: Inactive Devices](https://blog.jabberhead.tk/2019/12/13/pitfalls-for-omemo-implementations-part-1-inactive-devices/)<sup>(ext)</sup>
- [XEP-0384: Add message counter logic to determine stale devices #709](https://github.com/xsf/xeps/pull/709)<sup>(ext)</sup>

The original "OMEMO" plugin uses constants to define the following:

```python
DEFAULT_PREKEY_AMOUNT = 100
MIN_PREKEY_AMOUNT = 80
SPK_ARCHIVE_TIME = 86400 * 15  # 15 Days
SPK_CYCLE_TIME = 86400         # 24 Hours
UNACKNOWLEDGED_COUNT = 300
```
*ref:* [XEP-0384 Key Exchange](https://xmpp.org/extensions/xep-0384.html#protocol-key_exchange)<sup>(ext)</sup> - [XEP-0384 Bundles](https://xmpp.org/extensions/xep-0384.html#bundles)<sup>(ext)</sup>

> Quote from the "OMEMO" plugin [CHANGELOG](https://dev.gajim.org/gajim/gajim-plugins/-/blob/master/omemo/CHANGELOG)<sup>(ext)</sup>:
>
> ```
> 2.6.45 / 2019-02-21
> - Set device inactive after 300 unacknowledged messages
> ```

#### In summary

`UNACKNOWLEDGED_COUNT`, or plainly "unacknowledged message count", is the maximum number of messages, a threshold, after which a non-receiving device (device ID) will be marked as "inactive" and will no-further receive messages signed with its key.  

The rest of the constants meaning:  
- `DEFAULT_PREKEY_AMOUNT` is the number of PreKeys to provide in the bundle. The bundle SHOULD always contain around 100 PreKeys.
- `MIN_PREKEY_AMOUNT` it the minimum number of PreKeys to provide in the bundle. The bundle MUST always contain at least 25 PreKeys.
- `SPK_ARCHIVE_TIME` is a time period after which all existing SignedPreKeys will be deleted.
- `SPK_CYCLE_TIME` is the Signed PreKey rotation period. Signed PreKeys SHOULD be rotated periodically once a week to once a month. A faster or slower rotation period should not be required.

### Case Study - the Issue

Let's say you are using Gajim (same account) on more than one device.

Device_1 and Device_2 are not always turned-on or "online".

Initially everything works OK. OMEMO encrypted messages are signed for both devices and chat history between the two (or more) devices is being synced successfully.

An **issue** emerges when, for example, Device_1 gets turned-off or disconnected from the Internet (offline) and this state is preserved for a "long" time. *Defining "long" depends on user usage patterns. It can be hours, a day, a week or more.*

If you continue chatting on Device_2 and exchange more than 300 messages, non-receiving Device_1 will be marked as "inactive".

When Device_1 gets turned-on/connected again, Gajim will sync-down the first 300 signed OMEMO encrypted messages and after that, it will fill your chat-log with sequential warnings:  
`00:00:00 Nickname: This message was encrypted with OMEMO, but not for your device.`

While this behaviour is not a bug or an unexpected result, the real **issue** is what can someone do if one wants or needs to invalidate non-receiving devices after 600 messages or after just 1.

### Notes - on the Issue

["Conversations"](https://conversations.im/)<sup>(ext)</sup> app, for example, had a similar issue, [Allow user to configure how long to wait before making OMEMO user inactive #3499](https://github.com/iNPUTmice/Conversations/issues/3499)<sup>(ext)</sup> & [Extend the OMEMO auto expiry to 3 months](https://github.com/iNPUTmice/Conversations/pull/3584)<sup>(ext)</sup>.

"Conversations" is using a maximum time-period value before the invalidation of a device. A hard coded limit of 14 days was chosen, which it has now been extended to 3 months. The end user still has no access to change this, but a 3-month period is a reasonable amount of time.

A time-period based invalidation method of devices is maybe better than counting messages in "flexible security" situations (mainly for mobile devices).

A message count based invalidation method (Gajim) is probably "more secure", although inflexible, if your "inactive" device is stolen or lost. You certainly want the least amount of encrypted messages to be leaked.

### A Solution

*just an opinion*

Both applications, "Conversations" and Gajim in this particular case, chose not to let their aforementioned invalidation methods be user customizable and there definitely were good reasons to do so, considering modern mainstream usage patterns (?). Nevertheless, this "protective" choice is not practicable for everyone and every use-case.

Personally, while being neither a software engineer/developer nor a cryptographer, I think that the "unacknowledged message count" value should be left at the end user's discretion to alter it, according to their usage patterns and/or SecOps demands. The "OMEMO Mod" plugin provides this ability.


CHANGELOG
2.8.13.2 g1.4-dev Mod / 2022-04-29
- Sync with source (master branch - gajim-1.4.0-dev) v2.8.13.2 (g1.4-dev)
  - `UNACKNOWLEDGED_COUNT` default value to 2000
  - `UNACKNOWLEDGED_COUNT` max allowed value to 99999

2.8.12.2 g1.4-dev Mod / 2022-03-16
- Sync with source (master branch - gajim-1.4.0-dev) v2.8.12.2 (g1.4-dev)

2.7.15.2 Mod / 2021-11-04
- Sync with source v2.7.15

2.7.14.2.1 Mod / 2021-06-18
- input values validation allowing only digits

2.7.14.2 Mod / 2021-05-14
- Update the storage mechanism to manage all the "backend.util constants"
- Expose all "Advanced" OMEMO settings as user-configurable

2.7.14.1 Mod / 2021-05-11
- Storage mechanism for the UNACKNOWLEDGED_COUNT constant
- Expose UNACKNOWLEDGED_COUNT as a user configurable value

2.7.13 / 2021-01-08
- Fix file download

2.7.12 / 2020-12-06
- Adapt to Gajim changes

2.7.11 / 2020-10-19
- Adapt to nbxmpp changes

2.7.10 / 2020-10-11
- Adapt to nbxmpp changes

2.7.9 / 2020-09-19
- Adapt to Gajim changes

2.7.8 / 2020-09-12
- Remove not needed code

2.7.7 / 2020-06-28
- Add hints for the fingerprint dialog

2.7.6 / 2020-06-24
- Implement Blind Trust Before Verification
- Show outgoing messages as VERIFIED

2.7.5 / 2020-06-21
- Don't drop message silently on decryption error
- Don't include inactive devices on checking undecided trust

2.7.4 / 2020-05-23
- Sanitize BLOBs in database

2.7.3 / 2020-05-21
- Set sqlite text_factory to str

2.7.2 / 2020-05-21
- Use PEP-440 compatible version

2.7.1 / 2020-04-30
- Don't use deprecated attribute
- Use nbxmpp.namespaces

2.6.75 / 2020-04-04
- Adapt updating caps

2.6.74 / 2020-02-14
- Switch to 12 byte IV

2.6.73 / 2020-01-19
- Don't show warning for Key Transport messages

2.6.72 / 2020-01-17
- Add back 12 byte IV read support for aesgcm links

2.6.71 / 2019-12-29
- Refactor file downloads

2.6.70 / 2019-12-18
- Adapt to Gajim HTTPUpload changes

2.6.69 / 2019-12-06
- Adapt to nbxmpp changes

2.6.68 / 2019-11-18
- Correctly handle presence in anonymous rooms
- Fix error in config dialog if no account is available

2.6.67 / 2019-11-10
- Correctly handle MAM Messages in anonymous rooms
- Show error when message was not encrypted for our device
- Be consistent about omemo capable rooms

2.6.66 / 2019-10-05
- Adapt to nbxmpp changes

2.6.65 / 2019-08-21
- Fix printing status message

2.6.64 / 2019-08-20
- Set trust per JID

2.6.63 / 2019-08-19
- Fix ConfirmationDialog

2.6.62 / 2019-07-09
- Remove YesNoDialog

2.6.61 / 2019-06-30
- Dont validate JID on affiliation result

2.6.60 / 2019-06-22
- Adapt to Gajim changes

2.6.59 / 2019-04-27
- Adapt to link handler changes

2.6.58 / 2019-03-31
- Fix group chat messages

2.6.57 / 2019-03-26
- Refactoring

2.6.56 / 2019-03-25
- Fix database version

2.6.55 / 2019-03-24
- Fix database migration

2.6.54 / 2019-03-24
- Fix database bug with old values

2.6.53 / 2019-03-20
- Pass trust to Gajim
- Bug Fix

2.6.52 / 2019-03-11
- Bug Fix

2.6.51 / 2019-03-08
- Fix endless devicelist update loop
- Build sessions immediately
- Dont allow sending message while having undecided fingerprints
- Improve logging

2.6.50 / 2019-02-24
- Add support for key transport messages
- Bug Fixes

2.6.49 / 2019-02-24
- GUI Improvements for Key Dialog

2.6.48 / 2019-02-22
- Bug Fixes

2.6.47 / 2019-02-22
- Bug Fixes

2.6.46 / 2019-02-22
- Bug Fixes

2.6.45 / 2019-02-21
- Set device inactive after 300 unacknowledged messages
- Small refactoring
- Fix a bug where the KeyDialog could not be opened

2.6.44 / 2019-02-18
- Fix error while loading keys from database

2.6.43 / 2019-02-17
- Refactor

2.6.42 / 2019-02-13
- Refactor

2.6.41 / 2019-02-11
- Refactor

2.6.40 / 2019-01-05
- Fix label color
- Add fallback body

2.6.39 / 2019-01-04
- Adapt to Gajim changes

2.6.38 / 2019-01-01
- Add new devicelist module
- Fix deprecation warnings
- Fix an error in non-anonymous MUCs

2.6.37 / 2019-01-01
- Adapt to Gajim and nbxmpp changes

2.6.36 / 2018-12-22
- Fix carbon messages not working

2.6.35 / 2018-12-15
- Better handle key exchange messages

2.6.34 / 2018-11-20
- Use "current" as pubsub node id

2.6.33 / 2018-10-17
- Bug fix

2.6.32 / 2018-10-13
- Bug fix

2.6.31 / 2018-10-12
- GUI adjustments
- Use new fingerprints dialog
- Refactor some UI code

2.6.1 / 2018-08-08
- Save encryption details to the database
- Bugfixes

2.6.0 / 2018-08-08
- FingerprintDialog: show all fingerprints in one list

2.5.14 / 2018-07-15
- Make preparations for future Gajim versions
- Allow writing encrypted messages in groupchats if there are no other participants

2.5.14 / 2018-07-15
- Make preparations for future Gajim versions
- Allow writing encrypted messages in groupchats if there are no other participants

2.5.13 / 2018-05-21
- Bug fix

2.5.12 / 2018-05-20
- Bug fix

2.5.11 / 2018-04-25
- Bug fix

2.5.10 / 2018-04-09
- Bug fix

2.5.9 / 2018-04-08
- Remove pycrypo dependency

2.5.8 / 2018-03-31
- Bug fix

2.5.7 / 2018-02-26
- Big fix

2.5.6 / 2018-01-26
- Fix decrypting MAM MUC messages

2.5.5 / 2017-12-17
- Bug fix

2.5.4 / 2017-12-16
- Bug fix

2.5.3 / 2017-12-10
- Bug fix

2.5.2 / 2017-12-10
- Small refactoring

2.5.1 / 2017-11-21
- Bug fix

2.5.0 / 2017-11-20
- Add MAM for MUC decryption

2.4.3 / 2017-11-15
- Use Gajim API to announce caps

2.4.2 / 2017-11-15
- Query devicelists for contacts where we have none, this makes us a bit more independent from PEP
- Fix encrypting in Groupchats
- Improve error messages

2.4.1 / 2017-11-12
- Bug fix

2.4.0 / 2017-11-10
- Refactoring

2.3.8 / 2017-10-07
- Bug Fixes

2.3.7 / 2017-08-26
- Query only the most recent PEP items

2.3.6 / 2017-08-21
- Adapt to Gaim being now a Package

2.3.5 / 2017-08-07
- Support 12bit IVs on httpupload files

2.3.4 / 2017-06-10
- Bugfixes
- Some Refactoring

2.3.3 / 2017-06-09
- Move encryption logic for files from the HTTPUploadPlugin to the OMEMOPlugin

2.3.2 / 2017-06-02
- Adapt to patches regarding LMC in Gajim

2.3.1 / 2017-05-23
- Bugfixes

2.3.0 / 2017-05-07
- Make plugin compatible with Gajims encryption API


backend/util.py
# Copyright (C) 2015 Bahtiar `kalkin-` Gadimov <bahtiar@gadimov.de>
#
# This file is part of OMEMO Gajim Plugin.
#
# OMEMO Gajim Plugin is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published
# by the Free Software Foundation; version 3 only.
#
# OMEMO Gajim Plugin is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with OMEMO Gajim Plugin. If not, see <http://www.gnu.org/licenses/>.

import pickle
from pathlib import Path
from gajim.common import configpaths

import binascii
import textwrap
from enum import IntEnum

from axolotl.identitykey import IdentityKey


class Manage_Constants():

    def __init__(self):
        conf_dir = Path(configpaths.get('PLUGINS_CONFIG_DIR'))
        self.conf_file = conf_dir / 'omemo_mod.pickle'
        if not Path(self.conf_file).is_file():
            self.var = {
                'DEFAULT_PREKEY_AMOUNT': 100,
                'MIN_PREKEY_AMOUNT': 80,
                'SPK_ARCHIVE_TIME': 86400 * 15,  # 15 Days
                'SPK_CYCLE_TIME': 86400,         # 24 Hours
                'UNACKNOWLEDGED_COUNT': 2000
            }
            with open(self.conf_file, 'wb') as f:
                pickle.dump(self.var, f)
        else:
            with open(self.conf_file, 'rb') as f:
                self.var = pickle.load(f)

    def save_constants(self):
        with open(self.conf_file, 'wb') as f:
            pickle.dump(self.var, f)


CONSTANTS = Manage_Constants()


class Trust(IntEnum):
    UNTRUSTED = 0
    VERIFIED = 1
    UNDECIDED = 2
    BLIND = 3


def get_fingerprint(identity_key, formatted=False):
    public_key = identity_key.getPublicKey().serialize()
    fingerprint = binascii.hexlify(public_key).decode()[2:]
    if not formatted:
        return fingerprint
    fplen = len(fingerprint)
    wordsize = fplen // 8
    buf = ''
    for w in range(0, fplen, wordsize):
        buf += '{0} '.format(fingerprint[w:w + wordsize])
    buf = textwrap.fill(buf, width=36)
    return buf.rstrip().upper()


class IdentityKeyExtended(IdentityKey):
    def __hash__(self):
        return hash(self.publicKey.serialize())

    def get_fingerprint(self, formatted=False):
        return get_fingerprint(self, formatted=formatted)


gtk/config.py
# Copyright (C) 2019 Philipp Hörist <philipp AT hoerist.com>
# Copyright (C) 2015 Bahtiar `kalkin-` Gadimov <bahtiar@gadimov.de>
# Copyright (C) 2015 Daniel Gultsch <daniel@cgultsch.de>
#
# This file is part of OMEMO Gajim Plugin.
#
# OMEMO Gajim Plugin is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published
# by the Free Software Foundation; version 3 only.
#
# OMEMO Gajim Plugin is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with OMEMO Gajim Plugin. If not, see <http://www.gnu.org/licenses/>.

import logging

from gi.repository import Gtk
from gi.repository import Gdk

from gajim.common import app
from gajim.plugins.helpers import get_builder
from gajim.plugins.plugins_i18n import _

from omemo.backend.util import get_fingerprint, CONSTANTS

log = logging.getLogger('gajim.p.omemo')


class OMEMOConfigDialog(Gtk.ApplicationWindow):
    def __init__(self, plugin, transient):
        Gtk.ApplicationWindow.__init__(self)
        self.set_application(app.app)
        self.set_show_menubar(False)
        self.set_title(_('OMEMO Settings'))
        self.set_transient_for(transient)
        self.set_default_size(400, 400)
        self.set_type_hint(Gdk.WindowTypeHint.DIALOG)
        self.set_destroy_with_parent(True)

        self._plugin = plugin

        path = self._plugin.local_file_path('gtk/config.ui')
        self._ui = get_builder(path)

        image_path = self._plugin.local_file_path('omemo.png')
        self._ui.image.set_from_file(image_path)

        try:
            self.disabled_accounts = self._plugin.config['DISABLED_ACCOUNTS']
        except KeyError:
            self._plugin.config['DISABLED_ACCOUNTS'] = []
            self.disabled_accounts = self._plugin.config['DISABLED_ACCOUNTS']

        box = Gtk.Box()
        box.pack_start(self._ui.notebook1, True, True, 0)

        self.add(box)

        self._ui.connect_signals(self)
        self.show_all()

        self.plugin_active = False
        for plugin in app.plugin_manager.active_plugins:
            log.debug(type(plugin))
            if type(plugin).__name__ == 'OmemoPlugin':
                self.plugin_active = True
                break

        self.update_account_store()
        self.update_account_combobox()
        self.update_disabled_account_view()
        self.update_settings()

    def is_in_accountstore(self, account):
        for row in self._ui.account_store:
            if row[0] == account:
                return True
        return False

    def update_account_store(self):
        for account in sorted(app.settings.get_active_accounts()):
            if account in self.disabled_accounts:
                continue
            if account == 'Local':
                continue
            if not self.is_in_accountstore(account):
                self._ui.account_store.append(row=(account,))

    def update_account_combobox(self):
        if self.plugin_active is False:
            return
        if len(self._ui.account_store):
            self._ui.account_combobox.set_active(0)
        else:
            self.account_combobox_changed_cb(self._ui.account_combobox)

    def account_combobox_changed_cb(self, box, *args):
        self.update_context_list()

    def update_disabled_account_view(self):
        self._ui.disabled_account_store.clear()
        for account in self.disabled_accounts:
            self._ui.disabled_account_store.append(row=(account,))

    def activate_accounts_btn_clicked(self, _button, *args):
        selection = self._ui.disabled_accounts_view.get_selection()
        mod, paths = selection.get_selected_rows()
        for path in paths:
            it = mod.get_iter(path)
            account = mod.get(it, 0)
            if account[0] in self.disabled_accounts and \
                    not self.is_in_accountstore(account[0]):
                self._ui.account_store.append(row=(account[0],))
                self.disabled_accounts.remove(account[0])
        self.update_disabled_account_view()
        self._plugin.config['DISABLED_ACCOUNTS'] = self.disabled_accounts
        self.update_account_combobox()

    def disable_accounts_btn_clicked(self, _button, *args):
        selection = self._ui.active_accounts_view.get_selection()
        mod, paths = selection.get_selected_rows()
        for path in paths:
            it = mod.get_iter(path)
            account = mod.get(it, 0)
            if account[0] not in self.disabled_accounts and \
                    self.is_in_accountstore(account[0]):
                self.disabled_accounts.append(account[0])
                self._ui.account_store.remove(it)
        self.update_disabled_account_view()
        self._plugin.config['DISABLED_ACCOUNTS'] = self.disabled_accounts
        self.update_account_combobox()

    def cleardevice_button_clicked_cb(self, button, *args):
        active = self._ui.account_combobox.get_active()
        account = self._ui.account_store[active][0]
        app.connections[account].get_module('OMEMO').clear_devicelist()
        self.update_context_list()

    def refresh_button_clicked_cb(self, button, *args):
        self.update_context_list()

    def _on_blind_trust(self, button):
        self._plugin.config['BLIND_TRUST'] = button.get_active()

    def _alter_unacknowledged_count(self, button):
        if button.get_text().isdigit():
            datum = int(button.get_text()) \
                if button.get_text() != '' else 2000
            if datum == 0:
                datum = 1
        else:
            datum = 2000
        CONSTANTS.var['UNACKNOWLEDGED_COUNT'] = datum
        CONSTANTS.save_constants()

    def _alter_default_prekey_amount(self, button):
        if button.get_text().isdigit():
            datum = int(button.get_text()) \
                if button.get_text() != '' else 100
            if datum < 80:
                datum = 80
        else:
            datum = 100
        CONSTANTS.var['DEFAULT_PREKEY_AMOUNT'] = datum
        CONSTANTS.save_constants()

    def _alter_min_prekey_amount(self, button):
        if button.get_text().isdigit():
            datum = int(button.get_text()) \
                if button.get_text() != '' else 80
            if datum < 25:
                datum = 25
        else:
            datum = 80
        CONSTANTS.var['MIN_PREKEY_AMOUNT'] = datum
        CONSTANTS.save_constants()

    def _alter_spk_archive_time(self, button):
        if button.get_text().isdigit():
            datum = int(button.get_text()) \
                if button.get_text() != '' else 1296000
            if datum < 86400:
                datum = 86400
            elif datum > 5259600:
                datum = 5259600
        else:
            datum = 1296000
        CONSTANTS.var['SPK_ARCHIVE_TIME'] = datum
        CONSTANTS.save_constants()

    def _alter_spk_cycle_time(self, button):
        if button.get_text().isdigit():
            datum = int(button.get_text()) \
                if button.get_text() != '' else 86400
            if datum < 86400:
                datum = 86400
            elif datum > 2629800:
                datum = 2629800
        else:
            datum = 86400
        CONSTANTS.var['SPK_CYCLE_TIME'] = datum
        CONSTANTS.save_constants()

    def update_context_list(self):
        self._ui.deviceid_store.clear()

        if not len(self._ui.account_store):
            self._ui.ID.set_markup('')
            self._ui.fingerprint_label.set_markup('')
            self._ui.refresh.set_sensitive(False)
            self._ui.cleardevice_button.set_sensitive(False)
            return
        active = self._ui.account_combobox.get_active()
        account = self._ui.account_store[active][0]

        # Set buttons active
        self._ui.refresh.set_sensitive(True)
        if account == 'Local':
            self._ui.cleardevice_button.set_sensitive(False)
        else:
            self._ui.cleardevice_button.set_sensitive(True)

        # Set FPR Label and DeviceID
        omemo = self._plugin.get_omemo(account)
        self._ui.ID.set_markup('<tt>%s</tt>' % omemo.backend.own_device)

        identity_key = omemo.backend.storage.getIdentityKeyPair()
        fpr = get_fingerprint(identity_key, formatted=True)
        self._ui.fingerprint_label.set_markup('<tt>%s</tt>' % fpr)

        own_jid = app.get_jid_from_account(account)
        # Set Device ID List
        for item in omemo.backend.get_devices(own_jid):
            self._ui.deviceid_store.append([item])

    def update_settings(self):
        self._ui.blind_trust_checkbutton.set_active(
            self._plugin.config['BLIND_TRUST'])
        self._ui.unacknowledged_count.set_text(
            str(CONSTANTS.var['UNACKNOWLEDGED_COUNT']))
        self._ui.default_prekey_amount.set_text(
            str(CONSTANTS.var['DEFAULT_PREKEY_AMOUNT']))
        self._ui.min_prekey_amount.set_text(
            str(CONSTANTS.var['MIN_PREKEY_AMOUNT']))
        self._ui.spk_archive_time.set_text(
            str(CONSTANTS.var['SPK_ARCHIVE_TIME']))
        self._ui.spk_cycle_time.set_text(
            str(CONSTANTS.var['SPK_CYCLE_TIME']))
